File: python-backend/app/orchestrator/data_types.py
"""
Data type system for workflow node ports.
"""

PORT_TYPE_COMPATIBILITY = {
    "text": {"text", "any"},
    "json": {"json", "text", "any"},  # json can stringify to text
    "array": {"array", "json", "any"},
    "image": {"image", "any"},
    "number": {"number", "text", "any"},
    "boolean": {"boolean", "any"},
    "any": {"text", "json", "array", "image", "number", "boolean", "any"},
}


def is_compatible_connection(source_type: str, target_type: str) -> bool:
    """Check if source port type can connect to target port type."""
    return target_type in PORT_TYPE_COMPATIBILITY.get(source_type, set())

File: apps/web/client/src/lib/workflow/dataTypes.ts
/**
 * Data type system for workflow node connections.
 *
 * Defines the type system used for port compatibility validation.
 */

export type DataType = "text" | "json" | "array" | "image" | "number" | "boolean" | "any";

/**
 * Port type compatibility matrix.
 *
 * Maps each source data type to the target data types it can connect to.
 */
export const dataTypeCompatibility: Record<DataType, Set<DataType>> = {
  text: new Set<DataType>(["text", "any"]),
  json: new Set<DataType>(["json", "text", "any"]), // JSON can stringify to text
  array: new Set<DataType>(["array", "json", "any"]), // Arrays are JSON-compatible
  image: new Set<DataType>(["image", "any"]),
  number: new Set<DataType>(["number", "text", "any"]), // Numbers can convert to text
  boolean: new Set<DataType>(["boolean", "any"]), // Booleans can convert to text
  any: new Set<DataType>(["text", "json", "array", "image", "number", "boolean", "any"]), // Any accepts all
};

/**
 * Check if a connection between two port types is valid.
 *
 * @param sourceType - Data type of the source output port
 * @param targetType - Data type of the target input port
 * @returns true if the connection is valid, false otherwise
 */
export function isCompatibleConnection(sourceType: DataType, targetType: DataType): boolean {
  const compatibleTargets = dataTypeCompatibility[sourceType];
  return compatibleTargets ? compatibleTargets.has(targetType) : false;
}

/**
 * Get a human-readable description of a data type.
 */
export function getDataTypeDescription(dataType: DataType): string {
  const descriptions: Record<DataType, string> = {
    text: "Plain text or string value",
    json: "JSON object with structured data",
    array: "Array or list of items",
    image: "Image data (URL or binary)",
    number: "Numeric value",
    boolean: "True or false value",
    any: "Any data type (accepts all connections)",
  };

  return descriptions[dataType] || "Unknown data type";
}
